larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1173 @@
+    const double a = 0.19559;
+    const double m = 3.0;
+    const double X1 = 3.0;
+    const double X0 = 0.2000;
+    const double delta0 = 0.0;
+
+    if (X < X0)
+        return delta0;
+    else if ((X > X0) && (X < X1))
+        return 2 * X * std::log(10) - C + (a * (std::pow((X1 - X), m)));
+    else
+        return 2 * X * std::log(10) + C;
+}
+
+//----------------------------------------------------------------------------------------------------------------------------------
+
+double TrackDirectionTool::BetheBloch(double &T, double &M)
+{
+    const double K = 0.307075; // constant K in MeV cm mol^-1
+    const double z = 1;        // charge in e
+    const double Z = 18;       // Atomic number Z
+    const double A = 39.948;   // Atomic mass in g mol-1
+    const double m_e = 0.511;  // Mass of electron in MeV
+    const double rho = 1.396;  // Density of material in g cm^-3 (here: argon density)
+    const double I = 0.000188; // Ionisation energy in MeV
+
+    double p(std::sqrt((T * T) + 2 * T * M));
+    double gamma(std::sqrt(1 + ((p / M) * (p / M))));
+    double beta(std::sqrt(1 - 1 / (gamma * gamma)));
+
+    double T_max(2 * m_e * (beta * gamma * beta * gamma) / (1 + 2 * gamma * m_e / M + ((m_e / M) * (m_e / M))));
+    return rho * ((K * z * z * Z) / A) *
+           (0.5 * std::log(2 * m_e * T_max * (beta * gamma * beta * gamma) / (I * I)) - (beta * beta) - (0.5 * DensityCorrection(p, M))) /
+           (beta * beta); //in MeV/cm
+}
+//----------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::FillLookupTable(LookupTable &lookupTable, double M)
+{
+    std::map<int, double> lookupMap;
+    std::map<double, int> reverseLookupMap;
+
+    double currentEnergy(lookupTable.GetInitialEnergy()), binWidth(lookupTable.GetBinWidth());
+    int maxBin(0);
+
+    for (double n = 0; n < 100000; ++n)
+    {
+        double currentdEdx = BetheBloch(currentEnergy, M);
+
+        if ((currentdEdx * binWidth) >= currentEnergy)
+        {
+            double maxRange = (n * binWidth) + (currentEnergy / currentdEdx);
+            lookupTable.SetMaxRange(maxRange);
+            maxBin = n;
+
+            lookupMap.insert(std::pair<int, double>(n, 0.0));
+            reverseLookupMap.insert(std::pair<double, int>(0.0, n));
+            break;
+        }
+        else
+        {
+            lookupMap.insert(std::pair<int, double>(n, currentEnergy));
+            reverseLookupMap.insert(std::pair<double, int>(currentEnergy, n));
+        }
+
+        currentEnergy -= (currentdEdx * binWidth);
+    }
+
+    //remove redundant entries to make lookup much faster
+    for (std::map<int, double>::iterator it = lookupMap.begin(); it != lookupMap.end(); it++)
+    {
+        double n(it->first);
+        double val(it->second);
+        double distanceFromMaxRange((maxBin - n) * binWidth);
+
+        if (n == 0 || n == maxBin)
```

Comment:
Could check `n` immediately after assignment to avoid initialising `val` and `distanceFromMaxRange` if we don't care about this value for `n`?